diff -Nur shadow-19990307/configure.in shadow-19990307.pld/configure.in
--- shadow-19990307/configure.in	Sun Mar  7 19:04:34 1999
+++ shadow-19990307.pld/configure.in	Mon Jun  7 17:58:09 1999
@@ -162,21 +162,21 @@
 	fi
 done
 
-AC_MSG_CHECKING(location of utmp)
+AC_MSG_CHECKING(location of utmpx)
 for utmpdir in /var/run /var/adm /usr/adm /etc NONE; do
 	if test "$utmpdir" = "NONE"; then
-		AC_MSG_WARN(utmp file not found)
-	elif test -f $utmpdir/utmp; then
-		AC_DEFINE_UNQUOTED(_UTMP_FILE, "$utmpdir/utmp")
+		AC_MSG_WARN(utmpx file not found)
+	elif test -f $utmpdir/utmpx; then
+		AC_DEFINE_UNQUOTED(_UTMP_FILE, "$utmpdir/utmpx")
 		AC_MSG_RESULT($utmpdir)
 		break
 	fi
 done
 
-AC_MSG_CHECKING(location of faillog/lastlog/wtmp)
+AC_MSG_CHECKING(location of faillog/lastlog/wtmpx)
 for logdir in /var/log /var/adm /usr/adm /etc; do
 	if test -d $logdir; then
-		AC_DEFINE_UNQUOTED(_WTMP_FILE, "$logdir/wtmp")
+		AC_DEFINE_UNQUOTED(_WTMP_FILE, "$logdir/wtmpx")
 		AC_DEFINE_UNQUOTED(LASTLOG_FILE, "$logdir/lastlog")
 		AC_DEFINE_UNQUOTED(FAILLOG_FILE, "$logdir/faillog")
 		AC_MSG_RESULT($logdir)
