--- shadow-4.0.6/src/useradd.c	2004-11-18 21:45:00.713398344 +0100
+++ shadow-4.0.16/src/useradd.c	2006-06-08 01:17:05.580340031 +0300
@@ -72,10 +72,10 @@
 /*
  * These defaults are used if there is no defaults file.
  */
-static gid_t def_group = 100;
+static gid_t def_group = 1000;
 static const char *def_gname = "other";
-static const char *def_home = "/home";
-static const char *def_shell = "";
+static const char *def_home = "/home/users";
+static const char *def_shell = "/sbin/nologin";
 static const char *def_template = SKEL_DIR;
 static const char *def_create_mail_spool = "no";
 
@@ -89,7 +89,7 @@
 #define	VALID(s)	(strcspn (s, ":\n") == strlen (s))
 
 static const char *user_name = "";
-static const char *user_pass = "!";
+static const char *user_pass = "!!";
 static uid_t user_id;
 static gid_t user_gid;
 static const char *user_comment = "";
--- shadow-4.5/libmisc/find_new_gid.c~	2017-01-29 22:37:22.000000000 +0200
+++ shadow-4.5/libmisc/find_new_gid.c	2017-05-17 23:13:32.785253060 +0300
@@ -61,8 +61,8 @@
 		/* A requested ID is allowed to be below the autoselect range */
 		*preferred_min = (gid_t) 1;
 
-		/* Get the minimum ID range from login.defs or default to 101 */
-		*min_id = (gid_t) getdef_ulong ("SYS_GID_MIN", 101UL);
+		/* Get the minimum ID range from login.defs or default to 10 */
+		*min_id = (gid_t) getdef_ulong ("SYS_GID_MIN", 10UL);
 
 		/*
 		 * If SYS_GID_MAX is unspecified, we should assume it to be one
--- shadow-4.5/libmisc/find_new_uid.c~	2017-01-29 22:37:22.000000000 +0200
+++ shadow-4.5/libmisc/find_new_uid.c	2017-05-17 23:10:38.366687971 +0300
@@ -61,8 +61,8 @@
 		/* A requested ID is allowed to be below the autoselect range */
 		*preferred_min = (uid_t) 1;
 
-		/* Get the minimum ID range from login.defs or default to 101 */
-		*min_id = (uid_t) getdef_ulong ("SYS_UID_MIN", 101UL);
+		/* Get the minimum ID range from login.defs or default to 1 */
+		*min_id = (uid_t) getdef_ulong ("SYS_UID_MIN", 1UL);
 
 		/*
 		 * If SYS_UID_MAX is unspecified, we should assume it to be one
