--- shadow-4.0.4.1/src/userdel.c.orig	2003-12-17 10:43:30.000000000 +0100
+++ shadow-4.0.4.1/src/userdel.c	2004-02-24 01:15:34.489823544 +0100
@@ -810,9 +810,10 @@
 			pam_end (pamh, retval);
 	}
 
-	if (retval != PAM_SUCCESS)
+	if (retval != PAM_SUCCESS) {
 		fprintf (stderr, _("%s: PAM chauthtok failed\n"), Prog);
-	exit (1);
+		exit (1);
+	}
 
 	if (retval == PAM_SUCCESS)
 		pam_end (pamh, PAM_SUCCESS);
