diff -urN shadow-4.0.6.orig/configure.in shadow-4.0.6/configure.in
--- shadow-4.0.6.orig/configure.in	2004-11-18 15:47:10.912304224 +0100
+++ shadow-4.0.6/configure.in	2004-11-18 21:27:30.545048200 +0100
@@ -34,7 +34,7 @@
 
 AC_CHECK_FUNCS(a64l fchmod fchown fsync getgroups gethostname getspnam \
 	gettimeofday getusershell getutent initgroups lchown lckpwdf lstat \
-	memcpy memset setgroups sigaction strchr updwtmp updwtmpx)
+	memcpy memset setgroups sigaction strchr updwtmp updwtmpx setlocale)
 
 dnl Checks for typedefs, structures, and compiler characteristics.
 AC_C_CONST
