diff -Nur shadow-19990307/configure.in shadow-19990307.pld/configure.in
--- shadow-19990307/configure.in	Sun Mar  7 19:04:34 1999
+++ shadow-19990307.pld/configure.in	Mon Jun  7 17:58:09 1999
@@ -146,24 +146,24 @@
 done])
 AC_DEFINE_UNQUOTED(MAIL_SPOOL_FILE, "$shadow_cv_mailfile")
 
-AC_CACHE_CHECK([location of utmp], shadow_cv_utmpdir,
+AC_CACHE_CHECK([location of utmpx], shadow_cv_utmpdir,
 [for utmpdir in /var/run /var/adm /usr/adm /etc none; do
-	if test -f $shadow_cv_utmpdir/utmp; then
+	if test -f $shadow_cv_utmpdir/utmpx; then
 		break
 	fi
 done])
 if test "$shadow_cv_utmpdir" = "none"; then
-	AC_MSG_WARN(utmp file not found)
+	AC_MSG_WARN(utmpx file not found)
 fi
-AC_DEFINE_UNQUOTED(_UTMP_FILE, "$shadow_cv_utmpdir/utmp")
+AC_DEFINE_UNQUOTED(_UTMP_FILE, "$shadow_cv_utmpdir/utmpx")
 
-AC_CACHE_CHECK([location of faillog/lastlog/wtmp], shadow_cv_logdir,
+AC_CACHE_CHECK([location of faillog/lastlog/wtmpx], shadow_cv_logdir,
 [for shadow_cv_logdir in /var/log /var/adm /usr/adm /etc; do
 	if test -d $logdir; then
 		break
 	fi
 done])
-AC_DEFINE_UNQUOTED(_WTMP_FILE, "$shadow_cv_logdir/wtmp")
+AC_DEFINE_UNQUOTED(_WTMP_FILE, "$shadow_cv_logdir/wtmpx")
 AC_DEFINE_UNQUOTED(LASTLOG_FILE, "$shadow_cv_logdir/lastlog")
 AC_DEFINE_UNQUOTED(FAILLOG_FILE, "$shadow_cv_logdir/faillog")
 
