--- shadow-4.0.6/src/useradd.c	2004-11-18 21:45:00.713398344 +0100
+++ shadow-4.0.16/src/useradd.c	2006-06-08 01:17:05.580340031 +0300
@@ -72,10 +72,10 @@
 /*
  * These defaults are used if there is no defaults file.
  */
-static gid_t def_group = 100;
+static gid_t def_group = 1000;
 static const char *def_gname = "other";
-static const char *def_home = "/home";
-static const char *def_shell = "";
+static const char *def_home = "/home/users";
+static const char *def_shell = "/sbin/nologin";
 static const char *def_template = SKEL_DIR;
 static const char *def_create_mail_spool = "no";
 
@@ -89,7 +89,7 @@
 #define	VALID(s)	(strcspn (s, ":\n") == strlen (s))
 
 static const char *user_name = "";
-static const char *user_pass = "!";
+static const char *user_pass = "!!";
 static uid_t user_id;
 static gid_t user_gid;
 static const char *user_comment = "";
--- shadow-4.1.5.1/libmisc/find_new_gid.c~	2011-07-29 17:38:23.000000000 +0200
+++ shadow-4.1.5.1/libmisc/find_new_gid.c	2012-12-14 14:03:45.355638812 +0100
@@ -68,7 +68,7 @@
 			return -1;
 		}
 	} else {
-		gid_min = (gid_t) getdef_ulong ("SYS_GID_MIN", 101UL);
+		gid_min = (gid_t) getdef_ulong ("SYS_GID_MIN", 10UL);
 		gid_max = (gid_t) getdef_ulong ("GID_MIN", 1000UL) - 1;
 		gid_max = (gid_t) getdef_ulong ("SYS_GID_MAX", (unsigned long) gid_max);
 		if (gid_max < gid_min) {
--- shadow-4.1.5.1/libmisc/find_new_uid.c~	2011-07-29 17:39:16.000000000 +0200
+++ shadow-4.1.5.1/libmisc/find_new_uid.c	2012-12-14 14:03:29.645639367 +0100
@@ -68,7 +68,7 @@
 			return -1;
 		}
 	} else {
-		uid_min = (uid_t) getdef_ulong ("SYS_UID_MIN", 101UL);
+		uid_min = (uid_t) getdef_ulong ("SYS_UID_MIN", 1UL);
 		uid_max = (uid_t) getdef_ulong ("UID_MIN", 1000UL) - 1;
 		uid_max = (uid_t) getdef_ulong ("SYS_UID_MAX", (unsigned long) uid_max);
 		if (uid_max < uid_min) {
