--- shadow-4.0.6/configure.in.orig	2004-11-21 03:26:30.800445000 +0100
+++ shadow-4.0.6/configure.in	2004-11-21 04:32:23.842491656 +0100
@@ -151,7 +151,7 @@
 fi
 
 AC_CACHE_CHECK([location of utmp], shadow_cv_utmpdir,
-[for utmpdir in /var/run /var/adm /usr/adm /etc none; do
+[for shadow_cv_utmpdir in /var/run /var/adm /usr/adm /etc none; do
 	if test -f $shadow_cv_utmpdir/utmp; then
 		break
 	fi
